File: chapter_3_stacks_and_queues/python/3.4-queue_via_stacks.py
Stack = __import__('stack').Stack

# first in, first out


class MyQueue():
    def __init__(self):
        self.newest = Stack()
        self.oldest = Stack()

    def size(self):
        return self.newest.length + self.oldest.length

    def enqueue(self, value):
        self.newest.push(value)

    def dequeue(self):
        self.shift_stacks()
        return self.oldest.pop()

    def peek(self):
        self.shift_stacks()
        return self.oldest.peek()

    def shift_stacks(self):
        # If oldest is empty, push everything from newest to oldest
        if self.oldest.is_empty():
            while not self.newest.is_empty():
                self.oldest.push(self.newest.pop())


if __name__ == "main":
    q = MyQueue()
    for i in range(10):
        q.enqueue(i)
    print(q.newest.stack, q.oldest.stack)
    # [0, 1, 2, 3, 4, 5, 6, 7, 8, 9] []
    print(q.dequeue())  # 0
    print(q.newest.stack, q.oldest.stack)
    # [] [9, 8, 7, 6, 5, 4, 3, 2, 1]
    print(q.peek())  # 1
    print(q.size())  # 9

File: chapter_3_stacks_and_queues/js/3.4-queueViaStacks.js
const { Stack } = require("./stack")

/* first in, first out*/
class MyQueue {
  constructor() {
    this.newest = new Stack()
    this.oldest = new Stack()
  }

  size() {
    return this.newest.length + this.oldest.length
  }

  enqueue(value) {
    this.newest.push(value)
  }

  shiftStacks() {
    /* If oldest is empty, push everything from newest to oldest */
    if (this.oldest.isEmpty()) {
      while (!this.newest.isEmpty()) {
        this.oldest.push(this.newest.pop())
      }
    }
  }

  dequeue() {
    this.shiftStacks()
    return this.oldest.pop()
  }

  peek() {
    this.shiftStacks()
    return this.oldest.peek()
  }
}

const mq = new MyQueue()
const arr = [1, 2, 3, 4, 5]
for (let i of arr) mq.enqueue(i)
console.log(mq)
/*
MyQueue {
  newest: Stack { stack: [ 1, 2, 3, 4, 5 ], length: 5 },
  oldest: Stack { stack: [], length: 0 } }
*/
console.log(mq.dequeue()) // 1
console.log(mq)
/*
MyQueue {
  newest: Stack { stack: [], length: 0 },
  oldest: Stack { stack: [ 5, 4, 3, 2 ], length: 4 } }
*/
mq.enqueue(6)
console.log(mq)
/*
MyQueue {
  newest: Stack { stack: [ 6 ], length: 1 },
  oldest: Stack { stack: [ 5, 4, 3, 2 ], length: 4 } }
*/
console.log(mq.dequeue()) // 2
console.log(mq)
/*
MyQueue {
  newest: Stack { stack: [ 6 ], length: 1 },
  oldest: Stack { stack: [ 5, 4, 3 ], length: 3 } }
*/
console.log(mq.size()) // 4
console.log(mq.peek()) // 3
